### PR TITLE
Use `tracing` crate for logs

### PR DESCRIPTION
**Pull request by str4d**
This exposes any more structured `tracing` events that are logged within the Rust crates, while continuing to collect `log` records.

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/tests) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [ ] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the demo app and try the changes?
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after changes and when possible, leave it in a better place._
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the demo app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/ccbc1f602a6fb9668a96c0b9ce5e202d9e43aaca..35dc3cddc6d39bdc86f85f0196579084a37801f0) to fix some developer issues and lints.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/35dc3cddc6d39bdc86f85f0196579084a37801f0..7c358337738e786311778500fcfac775acb371ed) to fix some small bugs, and add API-conditional support for mapping `tracing` spans to [Android trace events](https://developer.android.com/topic/performance/tracing/custom-events-native).

**Comment by str4d** [comment]
While working on this PR, I noticed this:

https://github.com/zcash/zcash-android-wallet-sdk/blob/103d2748c64be79c396673736e9317422d5bd860/sdk-lib/src/main/java/cash/z/ecc/android/sdk/jni/RustBackend.kt#L354-L364

I (and I suspect the rest of core team) have been operating under the assumption that error messages etc. on the Rust side were being logged. I would like to have the log initialisation (which includes all of the changes in this PR) happen as soon as the Rust library is loaded, and then if there are verbosity issues related to debugging, we should fix those directly.

**Comment by str4d** [comment]
Aha, my memory is not wrong. Rust logs _were_ logged initially, but were disabled in ddf785e7d88b9bf062ce6b2476ec1004fa4aa253. What was the motivation for this?

**Comment by ccjernigan** [comment]
> Aha, my memory is not wrong. Rust logs _were_ logged initially, but were disabled in [ddf785e](https://github.com/zcash/zcash-android-wallet-sdk/commit/ddf785e7d88b9bf062ce6b2476ec1004fa4aa253). What was the motivation for this?

This pre-dates me.  Maybe @pacu knows?

**Review by ccjernigan — APPROVED** [review]

File: sdk-lib/src/main/rust/utils/trace.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                    metadata.name()
+                ),
+            },
+            None => error!("Span {:?} has no metadata", id),
+        });
+    }
+
+    fn on_exit(&self, _id: &span::Id, _ctx: Context<'_, S>) {
+        self.with_api(|api| {
+            unsafe { api.ATrace_endSection() };
```
Comment (daira) [inline sdk-lib/src/main/rust/utils/trace.rs:43]: If the span name had an internal NUL then `ATrace_endSection` will be called without a corresponding `ATrace_beginSection`. The [NDK docs](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1gacd849f3cdbe927673c0ad8740c369aed) say:
> `void ATrace_endSection()`
>
> Writes a tracing message to indicate that a given section of code has ended.
>
> This call must be preceeded by a corresponding call to [ATrace_beginSection](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1ga13768bfdb5d90166ec50bc41235e2797) on the same thread. Calling this method will mark the end of the most recently begun section of code, so care must be taken to ensure that [ATrace_beginSection](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1ga13768bfdb5d90166ec50bc41235e2797)/[ATrace_endSection](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1gacd849f3cdbe927673c0ad8740c369aed) pairs are properly nested and called from the same thread.

**Review by daira — APPROVED** [review]
utACK modulo the comment about `ATrace_beginSection`/`ATrace_endSection` nesting in an error case.

File: sdk-lib/src/main/rust/utils/trace.rs (hunk 1 above)
Comment (str4d) [reply sdk-lib/src/main/rust/utils/trace.rs:43]: There are similar issues if the Rust code's spans are across async Rust `.await` boundaries, which we can't deal with until API 29. I will be handling that in a separate PR.

**Pull request merged**